packages/transformers/js/core/src/collect.rs

```diff
@@ -689,6 +692,11 @@ impl Visit for Collect {
         return;
       }
       Expr::This(_this) => {
+        if self.is_module {
+          // Don't track this bindings if ESM syntax is present
```

Comment:
I think we might still have to track this_exprs in ESM (the second case below)? For example:

```js
export function a() {
  return this.b();
}

export function b() {
  return 2;
}
```

Looks like we only have a CJS test for that (in hoist.rs).

Reply:
This isn't valid code though as `this` is undefined for ESM files. Is there actual scenarios expecting this to work?

Reply:
if you do

```js
import * as ns from './above_code';

ns.a();
```

the `this` will be the namespace. Niklas brought this example up in the issue that the recent PR fixed. https://github.com/parcel-bundler/parcel/issues/7866#issuecomment-1079500173

Reply:
@devongovett I feel like this is an accidental implementation detail of ESM on top of existing JS. No one would do it in practice as it means you're library could only be used via namespace imports. Not sure we should go out of our way to support this.

Reply:
I only brought it up because I think this PR literally just fixed it! https://github.com/parcel-bundler/parcel/pull/9291 Does it work if you only make the first case conditional on ESM instead of the whole thing?

Reply:
@devongovett I'm a bit confused on this. That fix was for CJS files which still work. The ESM scenario doesn't work regardless of this change. If I add a test with the code Niklas mentioned in that PR it fails with and without my changes. 

> Does it work if you only make the first case conditional on ESM instead of the whole thing?

It does work in this case. I'll update the code to this but I'm not sure tracking `this` in ESM files is really doing anything.